src/lang.rs

```diff
@@ -1,10 +1,247 @@
 use crate::data;
+use annotate_snippets::snippet::{Annotation, AnnotationType, Slice, Snippet, SourceAnnotation};
 use nom;
 use nom::types::CompleteStr;
 use nom::*;
 use nom::{digit1, double, is_alphabetic, is_alphanumeric, is_digit, multispace};
+use nom_locate::LocatedSpan;
+use num_traits::FromPrimitive;
+use std::convert::From;
 use std::str;
 
+/// Wraps the result of the child parser in a Positioned and sets the start_pos and end_pos
+/// accordingly.
+macro_rules! with_pos {
+  ($i:expr, $submac:ident!( $($args:tt)* )) => ({
+      let start_pos: QueryPosition = $i.into();
+      match $submac!($i, $($args)*) {
+          Ok((i,o)) => Ok((i, Positioned {
+              start_pos,
+              value: o,
+              end_pos: i.into(),
+          })),
+          Err(e) => Err(e),
+      }
+  });
+  ($i:expr, $f:expr) => (
+    with_pos!($i, call!($f));
+  );
+}
+
+/// Type used to track the current fragment being parsed and its location in the original input.
+pub type Span<'a> = LocatedSpan<CompleteStr<'a>>;
+
+/// Container for the position of some syntax in the input string.  This is similar to the Span,
+/// but it only contains the offset.
+#[derive(Debug, PartialEq, Clone)]
+pub struct QueryPosition(pub usize);
+
+impl<'a> From<Span<'a>> for QueryPosition {
+    fn from(located_span: Span<'a>) -> Self {
+        QueryPosition(located_span.offset)
+    }
+}
+
+/// Container for values from the query that records the location in the query string.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Positioned<T> {
+    pub start_pos: QueryPosition,
+    pub end_pos: QueryPosition,
+    pub value: T,
+}
+
+impl<T> Positioned<T> {
+    pub fn into(&self) -> &T {
+        &self.value
+    }
+}
+
+/// Common syntax errors.
+#[derive(PartialEq, Debug, FromPrimitive, Fail)]
+pub enum SyntaxErrors {
+    #[fail(display = "")]
+    DelimiterStart,
+    #[fail(display = "unterminated double quote string")]
+    UnterminatedString,
+    #[fail(display = "expecting close parentheses")]
+    MissingParen,
+}
+
+impl SyntaxErrors {
+    fn to_resolution(&self) -> &'static str {
```

Comment:
should this be `Option<str>`? Not sure what happens when we say `add_resolution` and the resolution is empty